File: Data Structures & Algorithms/top-k-elements-in-list/submission-7.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = defaultdict(int)
        for num in nums:
            count[num] += 1
        
        res = [[] for _ in range(len(nums) + 1)]

        for key, val in count.items():
            res[val].append(key)

        output = []
        # [[], [1], [2], [3]]
        for i in range(len(res) - 1, 0, -1):
            arr = res[i]
            for num in arr:
                output.append(num)
                k -= 1
                if k == 0:
                    return output
        

        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-7.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int[] pre = new int[len], post = new int[len], res = new int[len];

        for(int i=0;i<len;i++){
            pre[i] = 1;
            post[i] = 1;
        }

        for (int i=1;i<len;i++){
            pre[i] = pre[i-1] * nums[i-1];
        }

        for (int i=len-2;i>-1;i--){
            post[i] = post[i+1] * nums[i+1];
        }

        for (int i=0;i<len;i++){
            res[i] = pre[i] * post[i];
        }

        return res;
        
    }
}  
